libraries/lib-numeric-formats/formatters/BeatsNumericConverterFormatter.cpp

```diff
@@ -0,0 +1,234 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/**********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file BeatsNumericConverterFormatter.cpp
+
+ Dmitry Vedenko
+
+ **********************************************************************/
+#include "BeatsNumericConverterFormatter.h"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+
+#include "NumericConverterRegistry.h"
+
+#include "SampleCount.h"
+
+namespace
+{
+
+class BeatsFormatter final :
+    public NumericConverterFormatter
+{
+public:
+   BeatsFormatter(
+      double tempo, int upperTimeSignature, int lowerTimeSignature,
+      int fracPart)
+       : mTempo { tempo }
+       , mUpperTimeSignature { upperTimeSignature }
+       , mLowerTimeSignature { lowerTimeSignature }
+       , mFracPart { fracPart }
+   {
+      mFields.reserve(mFracPart > 0 ? 3 : 2);
+
+      // Create "bars" field and digits. We don't really care about frac and base,
+      // Range is assumed to allow 999 bars. 
+      auto& barsField = mFields.emplace_back(NumericField { false, 1, 1000, true });
+      barsField.label = L" " + XO("bar").Translation() + L" ";
+
+      // Beats format is 1 based. For the time point "0" the expected output is "1 bar 1 beat [1]"
+      // For this reason we use (uts + 1) as the "range".
+      // On top of that, we want at least two digits to be shown. Audacity has a bug with the range of 11,
+      // where on digit will be calculated instead of 2.
+
+      auto& beatsField = mFields.emplace_back(
+         NumericField { false, 1, std::max(12, mUpperTimeSignature + 1), true });
+      beatsField.label = L" " + XO("beat").Translation();
+
+      if (mFracPart > 0)
+      {
+         beatsField.label += L" ";
+         // See the reasoning above about the range
+         auto& fracField = mFields.emplace_back(
+            NumericField { false, 1, std::max(12, mFracPart / mLowerTimeSignature + 1), true });
+      }
+
+      // Fill the aux mDigits structure
+      int pos = 0;
+      for (size_t i = 0; i < mFields.size(); i++)
+      {
+         mFields[i].pos = pos;
+
+         for (int j = 0; j < mFields[i].digits; j++)
+         {
+            mDigits.push_back(DigitInfo { int(i), j, pos });
+            pos++;
+         }
+
+         pos += mFields[i].label.length();
+      }
+
+      const auto beatLength = 60.0 / mTempo;
+      const auto barLength = mUpperTimeSignature * beatLength;
+      const auto fracLength = beatLength * mLowerTimeSignature / mFracPart;
```

Comment:
Didn't this cause you divide-by-zero crashes?

Reply:
There is a check mFracPart and there are no integer divisions.